File: library/src/main/java/mopel/io/library/BaseAnimator.java
package mopel.io.library;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.support.v4.view.ViewCompat;
import android.view.View;
import android.view.animation.Interpolator;

/**
 * Author: mopel
 * Date : 2017/4/23
 */

public abstract class BaseAnimator {
  private static final long DEFAULT_DURATION = 1000;

  private AnimatorSet animatorSet;
  private long duration = DEFAULT_DURATION;

  {
    animatorSet = new AnimatorSet();
  }

  protected abstract void prepare(View target);

  public void animate(){
    start();
  }

  public void restart(){
    animatorSet = animatorSet.clone();

  }

  public void start(){
    animatorSet.setDuration(duration);
    animatorSet.start();
  }

  public BaseAnimator setTarget(View target){
    reset(target);
    prepare(target);
    return this;
  }

  public void reset(View target){
    ViewCompat.setAlpha(target,1);
    ViewCompat.setScaleX(target,1);
    ViewCompat.setScaleY(target,1);
    ViewCompat.setTranslationX(target,1);
    ViewCompat.setTranslationY(target,1);
    ViewCompat.setTranslationZ(target,1);
    ViewCompat.setRotation(target,1);
    ViewCompat.setRotationX(target,1);
    ViewCompat.setRotationY(target,1);
  }

  public void cancel() {
    animatorSet.cancel();
  }

  public boolean isRunning() {
    return animatorSet.isRunning();
  }

  public boolean isStarted() {
    return animatorSet.isStarted();
  }


  public BaseAnimator setDuration(long duration) {
    this.duration = duration;
    return this;
  }

  public BaseAnimator setStartDelay(long delay){
    animatorSet.setStartDelay(delay);
    return this;
  }

  public BaseAnimator addAnimatorListener(Animator.AnimatorListener listener){
    animatorSet.addListener(listener);
    return this;
  }

  public BaseAnimator setInterpolator(Interpolator interpolator) {
    animatorSet.setInterpolator(interpolator);
    return this;
  }


  public void removeAnimatorListener(Animator.AnimatorListener l) {
    animatorSet.removeListener(l);
  }

  public void removeAllListener() {
    animatorSet.removeAllListeners();
  }


  public long getDuration() {
    return duration;
  }


  public AnimatorSet getAnimator() {
    return animatorSet;
  }
}

File: lib-kotlin/src/main/java/mopel/io/lib_kotlin/BaseAnimator.kt
package mopel.io.lib_kotlin

import android.animation.Animator
import android.animation.AnimatorSet
import android.support.v4.view.ViewCompat
import android.view.View
import android.view.animation.Interpolator

/**
 * Author: mopel
 * Date : 2017/4/23
 */

abstract class BaseAnimator {

    var animator: AnimatorSet = AnimatorSet()
        private set
    private var duration = DEFAULT_DURATION

    protected abstract fun prepare(target: View)

    fun animate() {
        start()
    }

    fun restart() {
        animator = animator.clone()

    }

    fun start() {
        animator.duration = duration
        animator.start()
    }

    fun setTarget(target: View): BaseAnimator {
        reset(target)
        prepare(target)
        return this
    }

    fun reset(target: View) {
        ViewCompat.setAlpha(target, 1f)
        ViewCompat.setScaleX(target, 1f)
        ViewCompat.setScaleY(target, 1f)
        ViewCompat.setTranslationX(target, 1f)
        ViewCompat.setTranslationY(target, 1f)
        ViewCompat.setTranslationZ(target, 1f)
        ViewCompat.setRotation(target, 1f)
        ViewCompat.setRotationX(target, 1f)
        ViewCompat.setRotationY(target, 1f)
    }

    fun cancel() {
        animator.cancel()
    }

    val isRunning: Boolean
        get() = animator.isRunning

    val isStarted: Boolean
        get() = animator.isStarted

    fun setStartDelay(delay: Long): BaseAnimator {
        animator.startDelay = delay
        return this
    }

    fun addAnimatorListener(listener: Animator.AnimatorListener): BaseAnimator {
        animator.addListener(listener)
        return this
    }

    fun setInterpolator(interpolator: Interpolator): BaseAnimator {
        animator.interpolator = interpolator
        return this
    }

    fun removeAnimatorListener(l: Animator.AnimatorListener) {
        animator.removeListener(l)
    }

    fun removeAllListener() {
        animator.removeAllListeners()
    }

    fun getDuration(): Long {
        return duration
    }

    fun setDuration(duration: Long): BaseAnimator {
        this.duration = duration
        return this
    }

    companion object {
        private val DEFAULT_DURATION: Long = 1000
    }
}
